qiskit_dynamics/solvers/lanczos.py

```diff
@@ -0,0 +1,291 @@
+        q_i, projection, alpha_i, beta_i, idx = val
+
+        q_p = q_i[idx - 1, :]
+        beta_p = beta_i[idx - 1]
+
+        q_i = q_i.at[[idx], :].set(projection.T / beta_p)
+        projection = A @ q_i[idx, :]
+        alpha_i = alpha_i.at[idx].set(q_i[idx, :].conj().T @ projection)
+        projection = projection - alpha_i[idx] * q_i[idx, :] - beta_p * q_p
+        beta_i = beta_i.at[idx].set(jnp.linalg.norm(projection))
+
+        delta = q_i[idx, :].conj().T @ projection
+        projection = projection - delta * q_i[idx, :]
+        alpha_i = alpha_i.at[idx].set(alpha_i[idx] + delta)
+
+        idx = idx + 1
+
+        val_next = [q_i, projection, alpha_i, beta_i, idx]
+        return val_next
+
+    (q_basis, _, alpha, beta, _) = while_loop(lanczos_iter_cond, lanczos_iter, initial_val)
+
+    tridiagonal = (
+        jnp.diag(alpha, k=0) + jnp.diag(beta[: k_dim - 1], k=-1) + jnp.diag(beta[: k_dim - 1], k=1)
+    )
+    q_basis = q_basis.T
+    return tridiagonal, q_basis
+
+
+@requires_backend("jax")
+def jax_lanczos_eigh(A: Array, y0: Array, k_dim: int):
+    """Finds the lowest (Algebraic) ``k_dim`` eigenvalues and corresponding eigenvectors of a
+    hermitian array using Lanczos algorithm. Implemented with ``jax``.
+    Args:
+        A : Array to diagonalize. Must be hermitian.
+        y0 : Vector to initialise Lanczos iteration.
+        k_dim : Dimension of the krylov subspace.
+
+    Returns:
+        q_basis : Basis of the krylov subspace.
+        eigen_values : lowest ``k_dim`` Eigenvalues.
+        eigen_vectors_t : Eigenvectors in krylov-space.
+    """
+
+    tridiagonal, q_basis = jax_lanczos_basis(A, y0, k_dim)
+    eigen_values, eigen_vectors_t = jnp.linalg.eigh(tridiagonal)
+
+    # Eigenvectors in hilbert-space.
+    # eigen_vectors_a = q_basis @ eigen_vectors_t
+
+    return q_basis, eigen_values, eigen_vectors_t
+
+
+@requires_backend("jax")
+def jax_lanczos_expm(
+    A: Array,
+    y0: Array,
+    k_dim: int,
+    scale_factor: Optional[float] = 1,
+):
+    """Calculates action of matrix exponential of an anti-hermitian array on the state using
+    Lanczos algorithm. Implemented with ``jax``.
+
+    Args:
+        A : Array to exponentiate. Must be anti-hermitian.
+        y0 : Initial state.
+        k_dim : Dimension of the krylov subspace.
+        scale_factor : Maximum step size.
+
+    Returns:
+        y_dt : Action of matrix exponential on state.
+
+    Raises:
+        ValueError : If ``y0`` is not 1d or 2d
+    """
+
+    if y0.ndim == 1:
+        A = 1j * A  # make hermitian
+        y0_norm = jnp.linalg.norm(y0)
+        q_basis, eigen_values, eigen_vectors_t = jax_lanczos_eigh(A, y0 / y0_norm, k_dim)
+        y_dt = (
+            q_basis
+            @ eigen_vectors_t
+            @ (jnp.exp(-1j * scale_factor * eigen_values) * eigen_vectors_t[0, :])
+        ) * y0_norm
+
+    elif y0.ndim == 2:
+        y_dt = [jax_lanczos_expm(A, yi, k_dim, scale_factor) for yi in y0.T]
```

Comment:
I think you either want to turn this loop into a `scan` or use `vmap`, to avoid long compile time issues coming from for loops with JAX.

Reply:
used scan instead of for-loop